opentelemetry-sdk/src/metrics/instrument.rs

```diff
@@ -242,18 +234,130 @@ impl Stream {
+    pub fn with_cardinality_limit(mut self, limit: usize) -> Self {
         self.cardinality_limit = Some(limit);
         self
     }
+
+    /// Build a new Stream instance using the configuration in this builder.
+    ///
+    /// # Returns
+    ///
+    /// A Result containing the new Stream instance or an error if the build failed.
+    pub fn build(self) -> Result<Stream, Box<dyn Error>> {
+        // TODO: Add same validation as already done while
+        // creating instruments. It is better to move validation logic
+        // to a common helper and call it from both places.
+        // The current implementations does a basic validation
+        // only to close the overall API design.
+
+        // if name is provided, it must not be empty
+        if let Some(name) = &self.name {
+            if name.is_empty() {
+                return Err("Stream name must not be empty".into());
+            }
+        }
+
+        // if cardinality limit is provided, it must be greater than 0
+        if let Some(limit) = self.cardinality_limit {
+            if limit == 0 {
+                return Err("Cardinality limit must be greater than 0".into());
+            }
+        }
+
+        // If the aggregation is set to ExplicitBucketHistogram, validate the bucket boundaries.
+        if let Some(Aggregation::ExplicitBucketHistogram { boundaries, .. }) = &self.aggregation {
+            validate_bucket_boundaries(boundaries)?;
+        }
+
+        Ok(Stream {
+            name: self.name,
+            description: self.description,
+            unit: self.unit,
+            aggregation: self.aggregation,
+            allowed_attribute_keys: self.allowed_attribute_keys,
+            cardinality_limit: self.cardinality_limit,
+        })
+    }
+}
+
+fn validate_bucket_boundaries(boundaries: &[f64]) -> Result<(), String> {
+    // Validate boundaries do not contain f64::NAN, f64::INFINITY, or f64::NEG_INFINITY
+    for boundary in boundaries {
+        if boundary.is_nan() || boundary.is_infinite() {
+            return Err(
+                "Bucket boundaries must not contain NaN, Infinity, or -Infinity".to_string(),
+            );
+        }
+    }
+
+    // validate that buckets are sorted and non-duplicate
+    for i in 1..boundaries.len() {
+        if boundaries[i] <= boundaries[i - 1] {
+            return Err("Bucket boundaries must be sorted and non-duplicate".to_string());
+        }
+    }
+
+    Ok(())
+}
+
+/// Describes the stream of data an instrument produces.
+///
+/// # Example
+///
+/// Streams can be used as masks in views.
+///
+/// ```
+/// use opentelemetry_sdk::metrics::{new_view, Aggregation, Instrument, Stream};
+///
+/// let criteria = Instrument::new().name("counter_*");
+/// let mask = Stream::builder()
+///     .with_aggregation(Aggregation::Sum)
+///     .build()
+///     .unwrap();
+///
+/// let view = new_view(criteria, mask);
+/// # drop(view);
+/// ```
+#[derive(Default, Debug)]
+#[non_exhaustive]
```

Comment:
can remove the non_exhaustive marker, as this is now a proper builder.?

Reply:
We can remove `non_exhaustive` mainly because a user cannot create a `Stream` struct as its fields are not public. The same goes for `StreamBuilder` struct as well.